compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if right == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(right),
+                            block_exit: right,
+                        });
+                    } else if right == *then_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: Some(right),
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        }
+    }
+    // A conditional exit would have exactly 2 predecessors
+    result.filter(|result| cfg.predecessors(result.block_exit).len() == 2)
+}
+
+/// Computes a cost estimate of a basic block
+/// returns u32::MAX if the block has side-effect instructions
+fn block_cost(block: BasicBlockId, dfg: &DataFlowGraph) -> u32 {
+    let mut cost: u32 = 0;
+
+    for instruction in dfg[block].instructions() {
+        let instruction_cost = match &dfg[*instruction] {
+            Instruction::Binary(binary) => {
+                match binary.operator {
+                    BinaryOp::Add { unchecked }
+                    | BinaryOp::Sub { unchecked }
+                    | BinaryOp::Mul { unchecked } => if unchecked { 1 } else { return u32::MAX },
+                    BinaryOp::Div
+                    | BinaryOp::Mod => return u32::MAX,
+                    BinaryOp::Eq => 2,
+                    BinaryOp::Lt => 5,
+                    BinaryOp::And
+                    | BinaryOp::Or
+                    | BinaryOp::Xor => 1,   //todo
+                    BinaryOp::Shl
+                    | BinaryOp::Shr => return u32::MAX,
+                }
+            },
+            Instruction::Cast(_, _) => 1,//TODO check if this instruction can fail
+            Instruction::Not(_) => 1,
+            Instruction::Truncate { .. } => 7,
+
+            Instruction::Constrain(_,_,_)
+            | Instruction::ConstrainNotEqual(_,_,_)
+            | Instruction::RangeCheck { .. }
+            | Instruction::Call { .. }  //TODO support calls with no-predicate set to true
+            |      Instruction::Load { .. }
+            | Instruction::Store { .. }
+            | Instruction::ArraySet { .. } => return u32::MAX,
+
+            Instruction::ArrayGet { array, index  } => {
+                // check if index is in bound
+                let index = dfg.get_numeric_constant(*index);
+                let mut ok_bound = false;
+                if let Some(index) = index {
```

Comment:
```suggestion
                let mut ok_bound = false;
                if let (Some(index), Some(len)) = (dfg.get_numeric_constant(*index), dfg.try_get_array_length(*array)) {
```
I think something like this would be more ergonomic and reduces nesting

Reply:
Done